### ISSUE TITLE
MVP of `[1603.45.16]` /"Ontologia"."United Nations"."P"/@eng-Latn

### ISSUE THREAD

**Issue by fititnt** [EticaAI/lexicographi-sine-finibus#2]
Quick links
- As part of _Dedicated repository to store examples of pre-compiled datasets #5_
- What are P-Codes
  - https://wiki.openstreetmap.org/wiki/P-Codes
  - https://en.wikipedia.org/wiki/Place_code
- Data source
  - https://data.humdata.org/dataset?ext_cod=1&res_format=XLSX
    -  https://drive.google.com/file/d/1jRshR0Mywd_w8r6W2njUFWv7oDVLgKQi/view?usp=sharing
- Related
  - HXL-CPLP/forum
    - https://github.com/HXL-CPLP/forum/issues/59#issuecomment-939403684 
  - _augusto-herrmann/transparencia-dados-abertos-brasil: Incluir mapeamento com códigos COD P-Code 26_
    - https://github.com/augusto-herrmann/transparencia-dados-abertos-brasil/issues/26

----

This issue is about minimal viable product of encode the entire public available P-Codes on numerordinatio. The scripts may need to get some cron job or manual upgrade over time, but this issue is mostly about at least have first version.

### Replacing ISO 3661-1 alpha 2 with UN M49
P-codes are prefixed with 2 letter codes, which have advantage of deal with leading zeros. So, for P-Codes, this make sense leading letters, which also allow use pure P-Codes as programming variables. However the numerordinatio works, we can go fully numeric.

### `[1603.45.16]` vs `[1603.45.49]`

In theory, `[1603.45.16]` **could** be a more specific version of `[1603.45.49]`  (https://unstats.un.org/unsd/methodology/m49/) instead of have own base namespace. This may change later.

Another point is that depending of how numerordinatio would be done, the codes could have aliases.

---

Changes
- `[1603.45.15]` renamed to `[1603.45.16]` (US-ASCII alphabet with K makes P as 16, not as 15).

**Comment by fititnt** [comment]
```bash
$ wc -l 1603/45/16/999/1603_45_16_1_15828996298662.hxl.csv 
432262 1603/45/16/999/1603_45_16_1_15828996298662.hxl.csv

m$ ls -lha  999999/1603/45/16/hxl/ | wc -l
408
$ ls -lha  999999/1603/45/16/hxl/*_0* | wc -l
114
$ ls -lha  999999/1603/45/16/hxl/* | wc -l
404
$ ls -lha  999999/1603/45/16/hxl/*0* | wc -l
114
$ ls -lha  999999/1603/45/16/hxl/*1* | wc -l
114
$ ls -lha  999999/1603/45/16/hxl/*2* | wc -l
110
$ ls -lha  999999/1603/45/16/hxl/*3* | wc -l
53
$ ls -lha  999999/1603/45/16/hxl/*4* | wc -l
13
$ ls -lha  999999/1603/45/16/hxl/*5* | wc -l
0

```
Trivia: do exist at least 432.262 published [Place codes](https://en.wikipedia.org/wiki/Place_code) worldwide. (from 0 to 4, not attested admin level 5 and 6).

The minimum an non-compressed CSV with every code would be around 13MB. Also, _how_ they are flattened would make difference on the space. But the good thing is we're far lower than GitHub ideal maximum of 50mb (hardlimit is 100MB)

### Conventions on how to use UN M49 private namespaces as reference for compiled results
On this topic

>> from wikipedia
>> https://en.wikipedia.org/wiki/UN_M49#Private-use_codes_and_reserved_codes
> Private-use codes and reserved codes
> Beside the codes standardized above, the numeric codes 900 to 999 are reserved for private-use in ISO 3166-1 (under agreement by the UNSD) and in the UN M.49 standard. They may be used for any other groupings or subdivision of countries, territories and regions.
> 
> Some of these private-use codes may be found in some UN statistics reports and databases, for their own specific purpose. They are not portable across databases from third parties (except through private agreement), and may be changed without notice.
> 
> Note that the code 000 is reserved and not used for defining any region. It is used in absence of data, or for data in which no region (not even the World as a whole) is applicable. For unknown or unencoded regions, private-use codes should preferably be used.

For aggregated datasets related to world places, I believe we should start using private namespaces for it **and document the logic**. This saves a lot of upfront drama with scripting.

### On logic about "population statistics"

I think aggregate population statistics is a different issue, but the sole major reason for the classical 70's UN m49 (https://unstats.un.org/unsd/publication/SeriesM/Series_M49_(1970)_en-fr.pdf) was this type of statistics. Wikipedia says this is not more used, but makes total sense for us here.

**However** I think population statistics is not a priority. But I know there is more than one datasets (and they are more automated) so at least at adm0 (country level) this would not be hard to automate. But we're already going for more detailed data, at least for countries such as Brazil which we may have additional sources.

Other priorities would be start mapping the P-Codes with Wikidata. Then things are going to be relevant.

**Comment by fititnt** [comment]
Related: 
- _**[praeparātiō ex automatīs] MVP of idea of automation to pre-process external data to LSF internal file format #42**_

---

Ok. We're gonna re-run the automation done in in January (the start of this repository) to know the file changes (if any) and then start to plan some way to try fetch regular updates (as per #42).

The type of content we're more interested are the P-Codes and ways to use them as reference for referential data for other information (such as population statistics) and eventually be able to interlink with references to Wikidata Q codes. The Brazilian ones (which we're doing on #40 and #42) will have far more details because Brazilian municipality code already exist (https://www.wikidata.org/wiki/Property:P1585), so thats one of the cases we get up to over 200 terminology translations (mostly popular cities).

### Modeling
Beyond HXL tags, since this is so already well structured, we can scale up some schema.

With the #41 (graph database, for now using RDF/SKOS) and #37 (relational database, like SQLite and PostgreSQL), we're both already able to do minimal viable products to ingest everything we can name on very structured numeric nomenclature in a way that no referential data will take the same database. This already works for relational database (https://github.com/EticaAI/lexicographi-sine-finibus/issues/37#issuecomment-1113897726), we just need to create a datapackage.json from frictionless, but RDF-like, **this will need to model everything**.

However, the way HXLTM and current tooling was more optimized to store linguistic data (so, values are stored as strings). But both RDF and databases already support gis-related storage. We can make people _accept_ the way to use numbers to reference dictionary endpoints and the codes themselves, but not only latitude as string, this would be a bit too much.

The humanitarian sector uses admin0, admin1, admin2, admin3, and I not sure if exist any ontology for this. Even if we _informally_ organize data, we could eventually try to get at least the nomenclature for these administrative divisions in a consistent way (which, again, would require reference point).

### Sources for production referential data (international level)
Our initial data was based on static dump on a zip file. The idea for #42 is make it on a regular basis, so we need to build over something else.

- **OCHA CODs (direct source)**
  - https://data.humdata.org/dashboards/cod by @OCHA-DAP et al
  - https://apps.itos.uga.edu/CODV2API by @UGA-ITOSHumanitarianGIS
- **CODs, indirect source**
  - https://fieldmaps.io/ by @maxmalynowsky
- **Other**
  - https://www.geoboundaries.org/ by @wmgeolab
    - This already is referenced by fieldmaps and HDX (except that I think are not CODs)

**Comment by fititnt** [comment]
> > TL;DR: **the way graph databases works means, in humanitarian jargon, a single 10mb to 200mb RDF file(*) can have the entire country-level data, but instead of Excel or SQL, users could use high level interfaces, such as Protege, to even have semantic reasoning**. While some humanitarians too focused would like the AI part of this, from our point of view this actually 1) helps to **allow the numeric codes we use be on local language of the persons who actually work on that at they own country"* and 2) **the initial very, very hard work to automate documentation eventually will use reasoning to validate itself"*.
> >
> > <sub>(*): however. the "single file" with all relevant to a region does not meant public data should be already such huge file because people would likely have own local data or want different public data. Either users adding each reference data or tools could merge the final result, but sensitive data do not need to leave users network.</sub>

----

# Rationale behind `[1603:16:{unm49}]` prefix

While we could start generating all the works weeks ago (and not going further on RDF SKOS) the after being viable load every dictionary we have to SQLite/PostgreSQL using well formed CSVs (topic #37), my time trying to "organize things" also in graph format (topic #41) made rethink the entire organization.

## under graph format, some entry points would have hunger amount of connected data

While there's different ways to partition data, unless we employ pure algorithmic (aka use some way to divide earth in equal blocks) most users will tend to organize or share data by administrative region. Humanitarian sector (at international level) tends to use more high levels, but places like Brazil would go much more specialized by region.

> > Side comment. Ok. After this, there's also cases were data would be shared by region (for example, coverage area of an hospital), which likely be more dynamic (and so, would need to have the area shared too), but for sake of both national and international interoperability, makes sense to make as easier as possible to everyone at bare minimum have ways to use the closest to standard codes.

One implication of the "way to organize in graph format" is that while something such as `[1603:45:16:76:2:3106200]` today we're using to represent the Belo Horizonte, both removing "45" is a number shorter and the amount of data attached to the `[1603:16:{unm49}]` very, very high. Actually, if we ignore translations for places names we can get from Wikidata (places such as Rio de Janeiro can have over 200), pretty much every time someone would use the namespaces with key by P codes, they will share final data, not data about CODs.

## On the idea that we might have other [1603:16:{unm49}] where "16" is a different nomenclature

After such simplification of remove the 45, as the entire idea is document/automate data interoperability (so, different from HXL Standard which only have vocabulary for aggregated, we're optimizing for > 90% use cases, e.g. sensitive data that should be processed offline) we will start to have namespaces for things that are not strictly places. For example:

- Natural persons
- Organizations

However, the way the world is organized, means **we would also plan ahead how to partition the data** which is not about place at all. Even if we consider Brazil alone, several government instances can have different entry points for same concept (like a person) and even at admin0 ("country" level in case of Brazil) for cases such as COVID-19 vaccines the entry points could up with public data go over 170 millions natural persons (>177.550.128). It's so much data that RDF wouldn't scale well compared to SQLite.

Ok. I'm aware it might sound "lazy" to recommend people to, unless someone else's in their country proposes better numeric namespace, as soon as we later create a base namespace to suggest as reference to store disaggregated data (at least Persons and Organizations) when in doubt, we suggest similar key to what we do to administrative boundaries.

This "lazy" way somewhat also helps to simplify a lot tooling that could allow (like by user additional parameter" to infer that such organization or person have its data coordinated by respectively administrative region. This might not be the case, but at least users have _some_ default suggestion that tools would work perfectly. In any case, it would still be possible to (at country level) different organizations have mappings of what one person or organization code inside them means on another level.

## A different "1603" on something like  `[1603:16:{unm49}]` might be used for data that is not 100% factual (like anonymized, test data, simulated data, etc)

This makes far more sense when dealing with data for natural persons or organizations, but we might have a totally different global prefix for data that is an entirely different class of data.

In a context of data directly associated with places [1603:16:{unm49}], this would means (at worst case to NOT use for real)

- The place country and it's internal divisions are a fake country; this might be relevant on simulations about epidemics we're people would complain if using real regions (less relevant if is country level preparedness; but not if other regions are invited to discuss)
- The places are real, but associated data is fully simulated. Since root 1603 would be different, tools would need humans to explicitly ask what data from "production data / reference data" can be applied to this simulated place.

This topic alone would require an open issue here about how we will handle this type of data. However there's some edge cases where what humanitarians use as production data already are "simulations" or statistical inferences based on real data (use case: country person-to-person Census is too old) so even something like persons living in an area might get quite complicated. For example, the same way we would "get data from more specific namespace" at [1603:45:16] (Humanitarian P-Codes) to publish on  [1603:16], since we're already preparing to allow ingest data on Tabular and Graph databases, we need somewhat assume users might not like our default choices, so things likely to have more versions (not just statistical Inference about population, which the entire methodology is different) makes sense have different options.

**Comment by fititnt** [comment]
Humm... we will need some documented way to 

1. Encode PCodes **by administrative level** which would also allow compatibility with RDF. 
    1. However, there is Wikidata P for UN M49, but not for P-Codes. (see https://www.wikidata.org/wiki/Wikidata:Database_reports/List_of_properties/all)
    2. We can still use https://en.wikipedia.org/wiki/Place_code which have the Q code https://www.wikidata.org/wiki/Q7200235 as something like `part_of`
3. For `#admX+i_nnn+i_Nnnn` without `+altN`, we can start encoding as `skos:prefLabel`
4. For `#admX+i_nnn+i_Nnnn` with `+altN`, we can start encoding as `skos:altLabel`, but since we can have several, it would be a good idea merge the fields with some separator

the current drafts already work for HXL / HXLTM, but without this change, it would need extra hardcoded logic. Another issue is that the current use of `+altN` allow pass validation of not having same fields (which `frictionless validate file.csv` and databases complain) but "is not semantic".

----

Edit: 
- a reason to encode PCodes **by administrative level** is for performance reasons when on graph databases. with over 450.000 P-Codes, if everyone become directly linked to a single upper concept, it would be not scale. Even dumping a Turtle file, it would get those > 450.000 potentially on a single line. Not good.
- We could still "merge" (maybe with RML) some way to allow a global search without administrative level, but at least this issue would be restricted to less queries. If the user or api generating the query at least know either the level OR that the prefix means a specific country, it could rewrite the query to much, much less search space. Either by administrative level OR prefix already would allow reduce drastically the performance issues (which makes sense only for cases where entire world is loaded, not really a issue for regional users)
